Refactor commit logic

alcaeus opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | 

#### Summary

[PHPORM-112](https://jira.mongodb.org/browse/PHPORM-112)

(This PR currently targets 2.6.x, but will be held back until we work on 2.7.x)

As is, failures during a `commit` operation in `UnitOfWork` result in an inconsistent state. While updates and deletions mostly handle errors during the operation gracefully, errors during an insert or upset result in the UnitOfWork having a broken state.

This PR changes the timing of when we mark changes as done and remove changesets. All of these are now handled after all commit operations have concluded, which is a change made in preparation for transaction support (which requires us to keep changesets around until we successfully commit the transaction or give up). As a side effect, this will eventually allow fixing the inconsistencies mentioned above. A new test class keeps track of these inconsistencies so they can be addressed in the future.

alcaeus committed:
Add tests for commit consistency showing wrong behaviour

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -147,26 +138,6 @@ public function addInsert(object $document): void
     }
 
-    /** @return array<string, object> */
-    public function getUpserts(): array
```
This class is marked as internal, so BC breaks shouldn't be an issue here. I don't think there's a benefit to keeping broken functionality around for the sake of backward compatibility.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -312,31 +254,16 @@ private function executeUpsert(object $document, array $options): void
          * NOOP otherwise. MongoDB 2.6+ does not allow empty modifiers, so $set
          * the identifier to the same value in our criteria.
          *
-         * This will fail for versions before MongoDB 2.6, which require an
-         * empty $set modifier. The best we can do (without attempting to check
-         * server versions in advance) is attempt the 2.6+ behavior and retry
-         * after the relevant exception.
-         *
```
We can safely remove this, as any supported driver version will refuse to connect to MongoDB 2.6 anyways. I don't see any benefits in keeping this around.

alcaeus commented on lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php:
```diff
@@ -312,31 +254,16 @@ private function executeUpsert(object $document, array $options): void
         assert($this->collection instanceof Collection);
-        $this->collection->updateOne($criteria, ['$set' => new stdClass()], $options);
+        $this->collection->updateOne($criteria, $data, $options);
+        $this->handleCollections($document, $options);
```
This was previously called in `executeUpserts`, so I moved it here. This is consistent with `update`, which also calls `handleCollections` itself.

alcaeus commented on lib/Doctrine/ODM/MongoDB/UnitOfWork.php:
```diff
@@ -444,41 +481,23 @@ public function commit(array $options = []): void
             $this->evm->dispatchEvent(Events::postFlush, new Event\PostFlushEventArgs($this->dm));
 
             // Clear up
-            $this->documentInsertions          =
-            $this->documentUpserts             =
-            $this->documentUpdates             =
-            $this->documentDeletions           =
```
These have been cleaned up in `cleanupExecutedChanges` and shouldn't be cleared entirely, as uncommitted changes may still be part of the UnitOfWork.

alcaeus commented on lib/Doctrine/ODM/MongoDB/UnitOfWork.php:
```diff
@@ -1187,13 +1217,8 @@ private function executeUpserts(ClassMetadata $class, array $documents, array $o
-
-        foreach ($documents as $document) {
+            $persister->upsert($document, $options);
+            $this->executedDocumentUpserts[] = $oid;
```
There is a subtle behavioural change here: while previously `postPersist` events were dispatched after all changes have been persisted, they are now dispatched after each individual change. This is consistent with `update` operation. I don't think it should post a big issue, but thought I'd point it out regardless.

alcaeus commented on lib/Doctrine/ODM/MongoDB/UnitOfWork.php:
```diff
@@ -1247,31 +1271,7 @@ private function executeDeletions(ClassMetadata $class, array $documents, array
-
-            // Document with this $oid after deletion treated as NEW, even if the $oid
-            // is obtained by a new document because the old one went out of scope.
-            $this->documentStates[$oid] = self::STATE_NEW;
```
Note: this has not been added in `cleanupExecutedChanges`, as unsetting the `documentState` works just as well. Any untracked entity is considered new unless otherwise indicated.

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -0,0 +1,197 @@
+use MongoDB\BSON\ObjectId;
+use Throwable;
+
+class UnitOfWorkCommitConsistencyTest extends BaseTestCase
```
These are the tests I used to test original behaviour and make sure that refactoring didn't break anything. I then used the undesired behaviour (see first commit in this PR) as a guideline for fixing the persister class.

GromNaN commented on lib/Doctrine/ODM/MongoDB/UnitOfWork.php:
```diff
@@ -162,42 +164,77 @@ final class UnitOfWork implements PropertyChangedListener
+     *
+     * @psalm-var array<string, PersistentCollectionInterface<array-key, object>>
+     */
+    private array $executedCollectionDeletions = [];
```
Listen to PHPStan which tells you that this property is useless.

alcaeus committed:
Clear scheduled document changes at the end of a commit operation

alcaeus committed:
Rename private variables to communicate intent

alcaeus commented:
After a number of unsuccessful attempts at getting non-transactional flushes to behave somewhat normal, I realised that this requires a lot more changes to UnitOfWork, DocumentPersister, and CollectionPersister than I'm willing to make at this point. I've kept the tests around, but changed the assertions to match current broken behaviour (which changes with this PR for some of the tests, but then you can't break something that isn't broken already).

Scheduled document changes and changesets are now cleared at the end of the commit operation only if the operation itself was successful. In other cases, no changes were made and a subsequent flush may want to persist data again. I'll try to address this once transactional commit is implemented, as that will give me a rough idea of how the behaviour differs.

jmikola commented on lib/Doctrine/ODM/MongoDB/UnitOfWork.php:
```diff
@@ -1161,9 +1161,8 @@ private function executeInserts(ClassMetadata $class, array $documents, array $o
-        foreach ($documents as $oid => $document) {
+        foreach ($documents as $document) {
             $persister->addInsert($document);
-            unset($this->documentInsertions[$oid]);
```
How does the removal of these `unset()` calls _during_ the execute methods accomplish the desired goal of clearing changes at the _end_ of a commit? Does this just rely on existing code in the commit step to wipe out all of the document HashTables (recently renamed with the "scheduled" prefix)?

alcaeus replied:
> Does this just rely on existing code in the commit step to wipe out all of the document HashTables

Correct. The initial approach was to track executed changes and clear only those that were actually executed in case the commit process fails with an exception. However, while this works fine for root documents, the handling of embedded documents in UnitOfWork needs to be reworked more significantly to allow for correctly marking embedded documents as handled, and correctly dispatching commit events for embedded documents is also more tricky.

For that reason, I've decided to not try and fix this broken UnitOfWork behaviour when a commit operation fails, and instead keep document changesets around until the commit is completed so that it can work with transactions.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -98,17 +97,17 @@ public function testErrorDuringInsertKeepsFailingInsertionsForDocument(): void
 
         // Wrong behaviour: user should still be scheduled for insertion
-        self::assertFalse($this->uow->isScheduledForInsert($secondUser));
+        self::assertTrue($this->uow->isScheduledForInsert($secondUser));
```
Why did the "Wrong behaviour" comment above pre-date the change from `assertFalse` to `assertTrue`?

alcaeus replied:
TL;DR: the comment is obsolete and was left in on error. I've removed it.

As for why the behaviour was (and still is) wrong, inserts and upserts are handled in batches: the operation is added to the persister, removed from the `scheduledDocumentInsertions` or `scheduledDocumentUpserts` hash table, then the persister executes all operations. This execution however is not all-or-nothing: for insertions, an `insertMany` is used, and for upserts each document is upserted separately using an `updateOne` operation. This meant that with the unique key violation in the second document the old behaviour was to mark all three documents as inserted (which was wrong), whereas now all three documents are still tracked as scheduled (which is still wrong).

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -465,8 +462,7 @@ public function testFailingDeleteWithEmbeddedDocumentKeepsChangeset(): void
         self::assertTrue($this->uow->isScheduledForDelete($user));
-        // Wrong behaviour: embedded document should be scheduled for deletion
-        self::assertFalse($this->uow->isScheduledForDelete($address));
+        self::assertTrue($this->uow->isScheduledForDelete($address));
```
Is this a legitimate bug fix or an unintended side effect? I assume it's related to having removed `$this->documentDeletions[$oid]` from `executeDeletes()`?

Same question applies to `isScheduledForInsert` and `isScheduledForUpsert` above, where you also removed "Wrong behaviour" comments.

alcaeus replied:
That's correct - unsetting this at the end fixes this behaviour but may introduce other inconsistencies. Commits without a transaction are a mess to track in UnitOfWork.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -0,0 +1,496 @@
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => [
+                'errorCode' => 10107, // Not primary
```
10107 is a [retryable write error](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#retryablewriteerror-labels). Do you actually need a second error to overcome the driver retrying by default, or are you explicitly disabling `retryWrites` when constructing the client?

alcaeus replied:
Ah, the "benefit" of testing with a standalone server. I've changed the error code to [192 (FailPointEnabled)](https://github.com/mongodb/mongo/blob/26be5a539d9e02519bbc17e1df6e44f8c3abbbfe/src/mongo/base/error_codes.yml#L238) for all fail points in this test. This makes more sense as we just want the operation to fail without a specific reason or remedy.

alcaeus committed:
Remove obsolete comment

alcaeus committed:
Use different error code

alcaeus committed:
Use single mongos in consistency tests

This ensures that the fail points are created on the same server that the write operations take place on, which can't be guaranteed in a sharded cluster with multiple mongoses.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/BaseTestCase.php:
```diff
@@ -162,4 +169,43 @@ protected function requireMongoDB42(string $message): void
+    {
+        $uri = getenv('DOCTRINE_MONGODB_SERVER') ?: DOCTRINE_MONGODB_SERVER;
+
+        return $useMultipleMongoses ? $uri : self::removeMultipleHosts($uri);
```
Expect this to fail horribly if we ever test with an SRV URI pointing to a sharded cluster.

alcaeus replied:
Yep, I'm hoping it won't come to that. Please point me to this comment with an appropriate gif once the time comes.

jmikola replied:
![](https://media.giphy.com/media/L17xM7PvLcqJggsCYa/giphy.gif)

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -0,0 +1,501 @@
+        self::assertTrue($this->uow->isScheduledForDelete($user));
+    }
+
+    public function testFailingDeleteWithEmbeddedDocumentKeepsChangeset(): void
```
Since this is derivative of the previous test, I think the names could be more similar:

 * testDeleteFailureKeepsChangeset
 * testDeleteFailureWithEmbeddedDocumentKeepsChangeset
 
Something to that effect.

alcaeus replied:
I've renamed all tests to be of the form `test<Operation>(Error)?<Description>`.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -0,0 +1,501 @@
+        self::assertTrue($this->uow->isScheduledForDelete($address));
+    }
+
+    public function testSuccessfulDeleteWithEmbeddedDocumentClearsChangeset(): void
```
Related to my previous comment, this could be:

 * testDeleteWithEmbeddedDocumentClearsChangeset

I don't think "Success" adds much value (it can be assumed).

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -0,0 +1,501 @@
+
+        $this->uow->remove($user);
+
+        // Make sure delete command fails once
+        $this->dm->getClient()->selectDatabase('admin')->command([
+            'configureFailPoint' => 'failCommand',
+            'mode' => ['times' => 1],
+            'data' => [
+                'errorCode' => 192, // FailPointEnabled
+                'failCommands' => ['delete'],
+            ],
+        ]);
```
Consider a private helper method that just takes a command name to remove all of this duplication and make the tests more readable.

alcaeus replied:
Good point.

jmikola commented on tests/Doctrine/ODM/MongoDB/Tests/UnitOfWorkCommitConsistencyTest.php:
```diff
@@ -0,0 +1,501 @@
+    {
+        // Create a unique index on the collection to let the second insert fail
+        $collection = $this->dm->getDocumentCollection(ForumUser::class);
+        $collection->createIndex(['username' => 1], ['unique' => true]);
```
Why bother with a unique index when `update` and `delete` use fail points?

alcaeus replied:
The `commit` call in this test will run `insertMany` with three documents. A fail point would make the `insert` command fail before writing the first document; however I specifically wanted to test the behaviour when the first document can be inserted but the second can't; the only way we can test this is by using an index error on the second document. I've expanded the comment.

jmikola reviewed (approved):
A few optional suggestions to improve the test file, but changes LGTM.

alcaeus committed:
Rename test methods for clarity

alcaeus committed:
Explain reasoning for index error

alcaeus committed:
Extract helper method to create failpoint
